Fix regression in TransactionTableModel

hebasto opened the pull request:
Since https://github.com/bitcoin/bitcoin/pull/17993 a crash is possible on exit.

Steps to reproduce:
- precondition: the old chain
- start `bitcoin-qt`
- wait until sync
- on main window: Menu -> File -> Quit
- crash

This PR is based on ryanofsky's [suggestion](https://github.com/bitcoin-core/gui/issues/7#issuecomment-646639251).

Fixes #7.

ryanofsky commented on src/qt/transactiontablemodel.cpp:
```diff
@@ -664,7 +664,7 @@ QVariant TransactionTableModel::headerData(int section, Qt::Orientation orientat
 {
     Q_UNUSED(parent);
-    TransactionRecord* data = priv->index(walletModel->wallet(), walletModel->clientModel().getBestBlockHash(), row);
+    TransactionRecord* data = priv->index(walletModel->wallet(), walletModel->getLastBlockProcessed(), row);
```
In commit "qt: Fix regression in TransactionTableModel" (cd3b6d0666fe0e6191596e2aa56980ab60841f2d)

Maybe also add null check on line 195 above:

https://github.com/bitcoin-core/gui/blob/cd3b6d0666fe0e6191596e2aa56980ab60841f2d/src/qt/transactiontablemodel.cpp#L195

```c++
if (!cur_block_hash.IsNull() && rec->statusUpdateNeeded(cur_block_hash) && wallet.tryGetTxStatus(rec->hash, wtx, numBlocks, block_time)) { 
```

To avoid spurious updateStatus calls during shutdown

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/8#issuecomment-646712144).

promag replied:
9735b0750bc1e510a5ca6764f9d3cdca5cc7f760

nit, assert block hash is not null in `TransactionRecord::statusUpdateNeeded`.

hebasto replied:
@promag
> nit, assert block hash is not null in `TransactionRecord::statusUpdateNeeded`.

Does this suggestion implies that `status.needsUpdate` should be the first operand of OR in `TransactionRecord::statusUpdateNeeded()`?

promag replied:
No, I don't mean to change that.

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/8#issuecomment-647760931).

hebasto commented:
Updated cd3b6d0666fe0e6191596e2aa56980ab60841f2d -> 9735b0750bc1e510a5ca6764f9d3cdca5cc7f760 ([pr8.01](https://github.com/hebasto/gui/commits/pr8.01) -> [pr8.02](https://github.com/hebasto/gui/commits/pr8.02), [diff](https://github.com/hebasto/gui/compare/pr8.01..pr8.02)):

- added @ryanofsky's [suggestion](https://github.com/bitcoin-core/gui/pull/8#discussion_r442886173)

vasild commented:
ACK 9735b075

Maybe both commits should be squashed because the second fixes an issue from the first.

promag commented on src/qt/walletmodel.h:
```diff
@@ -155,6 +155,9 @@ class WalletModel : public QObject
 
     void refresh(bool pk_hash_only = false);
+
+    uint256 getLastBlockProcessed();
```
cd3b6d0666fe0e6191596e2aa56980ab60841f2d

Can be const.

hebasto replied:
> Can be const.

Correct.

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/8#issuecomment-647760931).

promag commented:
I've a slight preference to "fix" before merging, suggestions so far are trivial.

hebasto commented:
@promag I will happy to fix, just awaiting a response for https://github.com/bitcoin-core/gui/pull/8#discussion_r443245815 :)

promag commented on src/qt/transactionrecord.cpp:
```diff
@@ -234,7 +234,7 @@ void TransactionRecord::updateStatus(const interfaces::WalletTxStatus& wtx, cons
 bool TransactionRecord::statusUpdateNeeded(const uint256& block_hash) const
 {
-    return status.m_cur_block_hash != block_hash || status.needsUpdate;
+    return (!block_hash.IsNull() && status.m_cur_block_hash != block_hash) || status.needsUpdate;
```
I meant
```cpp
assert(!block_hash.IsNull());
return status.m_cur_block_hash != block_hash || status.needsUpdate;
```

hebasto replied:
[Updated](https://github.com/bitcoin-core/gui/pull/8#issuecomment-647760931).

hebasto commented:
Updated 9735b0750bc1e510a5ca6764f9d3cdca5cc7f760 -> d906aaa117e337fc70575beecc0d6da314f57385 ([pr8.02](https://github.com/hebasto/gui/commits/pr8.02) -> [pr8.04](https://github.com/hebasto/gui/commits/pr8.04), [diff](https://github.com/hebasto/gui/compare/pr8.02..pr8.04)):

- addressed all comments
- squashed

hebasto commented:
@MarcoFalke Mind looking into this PR?
